### PR TITLE
nixos/netmeta: init

### PR DESCRIPTION
**Pull request by yu-re-ka**
###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by yu-re-ka** [comment]
We are now running this in production, does someone want to have another look?

File: pkgs/tools/networking/netmeta/default.nix [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    "cmd/helloworld"
+    "cmd/portmirror"
+    "cmd/reconciler"
+    "cmd/risinfo"
+  ];
+
+  vendorHash = "sha256-sgv3WhtqMcZoUYmBBBKhgNtpRYdW9dMWbB+4L3Gyv5E=";
+
+  meta = with lib; {
+    description = "scalable network observability toolkit optimized for performance";
```
Comment (Janik-Haag) [inline pkgs/tools/networking/netmeta/default.nix:27]: just a small nit
```suggestion
    description = "A scalable network observability toolkit optimized for performance";
```

File: pkgs/tools/networking/netmeta/default.nix [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+  src = fetchFromGitHub {
+    owner = "monogon-dev";
+    repo = "NetMeta";
+    rev = "af1f83e25e33c10fc6889c80d52b142a32d5521a";
+    hash = "sha256-T4OxUGgTbWWIo8W/4zZFHT415zR6KfDj3/fMitU19hc=";
+  };
+
+  subPackages = [
+    "cmd/helloworld"
```
Comment (Janik-Haag) [inline pkgs/tools/networking/netmeta/default.nix:18]: are you sure the helloworld package is needed? From what I can tell it only outputs a binary that repeatedly the current date/time and `I bims 1 container`

**Review by Janik-Haag — COMMENTED** [review]
Some feedback on the package/s, I only skimmed over the module (will take a closer look another time) but a test would be pretty awesome since it has many moving parts, if you don't want to write one that's also fine it's not a blocker.

**Comment by yu-re-ka** [comment]
Thanks for looking at my work 🙂 

> Perhaps it would make more sense for some of this to instead be maintained in the NetMeta repo, with a separate Cue target which only (lazily) renders the parts that the Nix deployment cares about?

The separate Cue target with a more defined interface would of course be very welcome, and then it can be consumed by the NixOS module. I don't see it as a requirement though.
I'm positive that I can figure out breakages of the interface with NetMeta soon enough, if and when it breaks. (It is really just taking single files out of k8s container descriptions.)

> It also relies on the nixpkgs versions of ClickHouse and goflow2 sufficiently matching the ones we ship, and while the stars currently happen to align perfectly, it is not guaranteed. For example, we make extensive use of ClickHouse features and ClickHouse frequently ships incompatible changes, new features and/or regressions.

I am very aware of the possible breakages that could arise in the future with dependency versions. Generally we will attempt to participate in upstream (goflow2, ClickHouse) discussion if there are _unexpected_ breakages.
As long as NetMeta does not fall more than a year behind any breaking ClickHouse changes, it should be doable.

Let's say hypothetically if there was a Nix flake in the NetMeta repository. Then there would be a pinned ClickHouse version packaged within the NetMeta repository too. This means security updates of the underlying ClickHouse (which contains a vendored openssl!) are not coming from the normal OS updates.
It's the question of viewing NetMeta as an appliance, vs seeing it as a program packaged for a distribution. There is no universal solution to this, since it is a tradeoff between getting OS updates and 'just works out of the box'.
In this case, I don't mind in putting some extra effort to achieve and maintain the integrated Distro approach, and I would argue other NixOS users care about this too because otherwise they would just press the 'deploy' button on the k8s based appliance.

**Review by leoluk — CHANGES_REQUESTED** [review]

File: pkgs/tools/networking/netmeta/default.nix [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    "cmd/risinfo"
+  ];
+
+  vendorHash = "sha256-sgv3WhtqMcZoUYmBBBKhgNtpRYdW9dMWbB+4L3Gyv5E=";
+
+  meta = with lib; {
+    description = "scalable network observability toolkit optimized for performance";
+    homepage = "https://github.com/monogon-dev/NetMeta";
+    license = licenses.asl20;
+    maintainers = with maintainers; [ yuka ];
```
[the remark this answers is not in the document]
Comment (yu-re-ka) [reply pkgs/tools/networking/netmeta/default.nix]: changed maintainers to wdz team

File: nixos/modules/services/networking/netmeta.nix [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+  images_local = mkCue "k8s" {
+    netmeta.images = {
+      helloworld = { image = ""; digest = ""; };
+      risinfo = { image = ""; digest = ""; };
+      goflow = { image = ""; digest = ""; };
+      portmirror = { image = ""; digest = ""; };
+      reconciler = { image = ""; digest = ""; };
+      grafana = { image = ""; digest = ""; };
+    };
+  };
```
[the remark this answers is not in the document]
Comment (yu-re-ka) [reply nixos/modules/services/networking/netmeta.nix:24]: The cue expressions do not evaluate without this

File: nixos/modules/services/networking/netmeta.nix [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+      clickhouseAdminPassword =  "";
+      clickhouseReadonlyPassword =  "";
+      sessionSecret = "<generate and paste secret here>";
+
+      publicHostname = "flowmon.example.com";
+
+      letsencryptMode = "staging"; # change to "production" later
+      letsencryptAccountMail = "letsencrypt@example.com";
+    } // cfg.settings;
+  };
```
[the remark this answers is not in the document]
Comment (yu-re-ka) [reply nixos/modules/services/networking/netmeta.nix:38]: Same here - The cue expressions do not evaluate without this. We specify example values to make the cue expression evaluate.

File: pkgs/tools/networking/goflow2/1.x.nix [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+  ldflags = [
+    "-s"
+    "-w"
+    "-X=main.version=${version}"
+  ];
+
+  vendorHash = "sha256-jWm/061alLKTzn53uKun1qj2TM77pjltQlZl1/dTd80=";
+
+  meta = with lib; {
```
Comment (NetaliDev) [inline pkgs/tools/networking/goflow2/1.x.nix:25]: `lib.getExe` in the module complains that `meta.mainProgram` is not set here.

**Review by johannwagner — APPROVED** [review]
I think, the packaging is as good as it gets.
We will maintain this mess, because we use this in multiple production scenarios. As @leoluk mentioned, there is some boilerplate but its needed for now to use netmeta without a huge patchset.

This PR introduces two new package versions, Clickhouse v23.3 and goflow2 v1. Both are currently supported and receive updates. We hope that we can remove those versions after updates of upstream.

File: pkgs/tools/networking/goflow2/1.x.nix (hunk 6 above)
Comment (yu-re-ka) [reply pkgs/tools/networking/goflow2/1.x.nix:25]: fixed

File: pkgs/servers/monitoring/grafana/plugins/netsage-sankey-panel/default.nix [hunk 7]
```diff
@@ -0,0 +1,7 @@
+{ grafanaPlugin }:
+
+grafanaPlugin {
+  pname = "netsage-sankey-panel";
+  version = "1.1.0";
+  zipHash = "sha256-cPBB9ksKhgnWbbo0C/vRHzsWk9Sh1hyh6h0AgwqaNKs=";
```
[the remark this answers is not in the document]
Comment (yu-re-ka) [reply pkgs/servers/monitoring/grafana/plugins/netsage-sankey-panel/default.nix]: Done

File: nixos/modules/services/networking/netmeta.nix [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+  options = with lib; {
+    services.netmeta = {
+      enable = mkEnableOption (lib.mdDoc "netmeta");
+
+      settings = mkOption {
+        type = types.attrs;
+        default = {};
+        description = lib.mdDoc ''
+          Additional settings for NetMeta.
+          Check [config.cue](https://github.com/monogon-dev/NetMeta/blob/8c5f861d89903d858b7266187e703f3d20ea4c07/deploy/single-node/config.cue) for all available settings.
```
[the remark this answers is not in the document]
Comment (yu-re-ka) [reply nixos/modules/services/networking/netmeta.nix:99]: Done

**Comment by yu-re-ka** [comment]
Let me summarize the state of this again:

As Johann described, we maintain the ClickHouse LTS branch for use with NetMeta. I plan to backport this to 23.11 with a warning on the ClickHouse LTS branch, that it is not supported for the entirety of the release cycle.
The assumption is that before the EOL of ClickHouse 23.3, NetMeta will have been updated to work with a newer branch. 
For users of the NetMeta module, the warning about ClickHouse 23.3 will be hidden, since we assume that they will receive a backport of an updated NetMeta version when it is ready.
In case NetMeta does not get updated in time for the ClickHouse 23.3 EOL, it would be marked as insecure.

Of course we would be very happy to work together with the upstream maintainers (feel free to contact me on matrix about this any time) to improve the interface between Nix(OS) and NetMeta, assist with building the container images using Nix, and building an improved single node deployment solution within the NetMeta repo. The in-tree NixOS module is a separate thing that is not in conflict with this, but we want it for the predictable security updates and not using a pinned nixpkgs, which is simply not possible with this much integration without an in-tree module.

File: pkgs/tools/networking/netmeta/default.nix (hunk 3 above)
[the remark this answers is not in the document]
Comment (yu-re-ka) [reply pkgs/tools/networking/netmeta/default.nix]: sorry, I don't know what to set as the mainProgram here, it is not really applicable

File: pkgs/servers/monitoring/grafana/plugins/netsage-sankey-panel/default.nix [hunk 9]
```diff
@@ -0,0 +1,15 @@
+{ grafanaPlugin }:
```
Comment (NetaliDev) [inline pkgs/servers/monitoring/grafana/plugins/netsage-sankey-panel/default.nix:1]: `lib` is missing here which, leads to a failing eval.

File: pkgs/servers/monitoring/grafana/plugins/netsage-sankey-panel/default.nix [hunk 10]
```diff
@@ -0,0 +1,15 @@
+{ lib, grafanaPlugin }:
+
+grafanaPlugin rec {
+  pname = "netsage-sankey-panel";
+  version = "1.1.1";
+  zipHash = "sha256-cPBB9ksKhgnWbbo0C/vRHzsWk9Sh1hyh6h0AgwqaNKb=";
```
Comment (NetaliDev) [inline pkgs/servers/monitoring/grafana/plugins/netsage-sankey-panel/default.nix:6]:
```
[ERROR]   stderr) error: hash mismatch in fixed-output derivation '/nix/store/1srq5dj1vzi7w6gimsiqc6i4wxz2skqf-netsage-sankey-panel-1.1.1.zip.drv':
[ERROR]   stderr)          specified: sha256-cPBB9ksKhgnWbbo0C/vRHzsWk9Sh1hyh6h0AgwqaNKY=
[ERROR]   stderr)             got:    sha256-WmXcIsV6jP2Z53qbqsEAElQI+VzoJFaVU8ib91+ZYM8=
```

**Review by NetaliDev — APPROVED** [review]
Tested it and it seems to work as it should.

**Comment by fionera** [comment]
> Of course we would be very happy to work together with the upstream maintainers (feel free to contact me on matrix about this any time)
Already done :smile: 

> to improve the interface between Nix(OS) and NetMeta, assist with building the container images using Nix, and building an improved single node deployment solution within the NetMeta repo. 
The current idea is to migrate NetMeta away from Bazel an build the containers with Nix. Additionally the application should be seperated from the deployment, e.g. a monorepo-ish setup with the application and then example deployments for nix, k3s, normal k8s, docker-compose or whatever floats your boat.

> The in-tree NixOS module is a separate thing that is not in conflict with this, but we want it for the predictable security updates and not using a pinned nixpkgs, which is simply not possible with this much integration without an in-tree module.
I see this issue, but I would also propose to wait for a refactor of NetMeta to have a proper interface. Currently there are lots of hacks that will break as soon as e.g. the Clickhouse deployment inside NetMeta changes.

**Comment by Janik-Haag** [comment]
> We would kindly request for this PR not be merged in the current state, or it be renamed to something like netmeta-lite, to avoid unnecessary work and migrations. We are already receiving support requests by people using this draft PR in production.

Gonna mark it as draft in order to prevent that.

**Pull request closed without merging**